crates/cheatnet/src/execution/cheatable_syscall_handler.rs

```diff
@@ -0,0 +1,297 @@
+    Ok(felt_to_stark_felt(&felt_from_ptr_immutable(vm, ptr)?))
+}
+
+pub fn felt_from_ptr_immutable(
+    vm: &VirtualMachine,
+    ptr: &Relocatable,
+) -> Result<Felt252, VirtualMachineError> {
+    let felt = vm.get_integer(*ptr)?.into_owned();
+    Ok(felt)
+}
+
+fn get_syscall_operand(hint: &StarknetHint) -> Result<&ResOperand, HintError> {
+    let StarknetHint::SystemCall { system: syscall } = hint else {
+        return Err(HintError::CustomHint(
+            "Test functions are unsupported on starknet.".into(),
+        ));
+    };
+    Ok(syscall)
+}
+
+fn get_syscall_cost(syscall_selector: SyscallSelector) -> u64 {
+    match syscall_selector {
+        SyscallSelector::LibraryCall | SyscallSelector::CallContract => {
+            constants::CALL_CONTRACT_GAS_COST
+        }
+        SyscallSelector::Deploy => constants::DEPLOY_GAS_COST,
+        SyscallSelector::GetExecutionInfo => constants::GET_EXECUTION_INFO_GAS_COST,
+        _ => unreachable!("Syscall has no associated cost"),
+    }
+}
+
+impl CheatableSyscallHandler<'_> {
+    fn execute_next_syscall_cheated(
+        &mut self,
+        vm: &mut VirtualMachine,
+        hint: &StarknetHint,
+    ) -> HintExecutionResult {
+        // We peek into the selector without incrementing the pointer as it is done later
+        let syscall = get_syscall_operand(hint)?;
+        let initial_syscall_ptr = get_ptr_from_res_operand_unchecked(vm, syscall);
+        let selector =
+            SyscallSelector::try_from(stark_felt_from_ptr_immutable(vm, &initial_syscall_ptr)?)?;
+        self.verify_syscall_ptr(initial_syscall_ptr)?;
+
+        match selector {
+            SyscallSelector::GetExecutionInfo => self.execute_syscall(
+                vm,
+                cheated_syscalls::get_execution_info_syscall,
+                SyscallSelector::GetExecutionInfo,
+            ),
+            SyscallSelector::CallContract => self.execute_syscall(
+                vm,
+                cheated_syscalls::call_contract_syscall,
+                SyscallSelector::CallContract,
+            ),
+            SyscallSelector::LibraryCall => self.execute_syscall(
+                vm,
+                cheated_syscalls::library_call_syscall,
+                SyscallSelector::LibraryCall,
+            ),
+            SyscallSelector::Deploy => self.execute_syscall(
+                vm,
+                cheated_syscalls::deploy_syscall,
+                SyscallSelector::Deploy,
+            ),
+            SyscallSelector::EmitEvent => {
+                let events_len_before = self.syscall_handler.events.len();
+                let hint_exec_result = self.syscall_handler.execute_next_syscall(vm, hint);
+                assert_eq!(
+                    events_len_before + 1,
+                    self.syscall_handler.events.len(),
+                    "EmitEvent syscall is expected to emit exactly one event"
+                );
+                cheated_syscalls::emit_event_hook(self);
```

Comment:
It is more of a syscall_hook cheated_syscall, it doesn't modify syscall logic itselt